clac.1
.Dd May 3, 2017
.Dt CLAC 1
.
.Sh NAME
.
.Nm clac
.Nd A command line, stack-based calculator with postfix notation

.Sh SYNOPSIS
.
.Nm
.Op Ar expression
.
.Sh DESCRIPTION
.
.Nm
is a command line, stack-based calculator with postfix notation
that displays the stack contents at all times. As you type, the
stack changes are reflected immediately.
.Pp
In a stack-based postfix calculator, entering a number pushes it
on a stack, and arithmetic operations pop their arguments from the
stack and push the result. As all the operations take a fixed number
of arguments, there's no room for ambiguity: parentheses and operator
precedence are not needed. Postfix notation is also known as reverse
Polish notation, or RPN.
.
.Ss Non-interactive mode
.
If an
.Em expression
is provided, clac will process it and print each element in the
stack starting from the top. It will then exit immediately.
.
.Ss Commands
.
When a command requires an argument, it pops a value from the stack.
If the stack is empty, a zero is provided instead. In the descriptions
below, the top of the stack (and thus the first value popped) is
represented by the letter `a`, while the second value popped is
represented by the letter `b`. For example, if the stack is composed
of the number `1`, `2` and `3` (with `3` at the top of the stack),
when we describe the sum then `a` will be `3` and `b` will be `2`.
It is important to note that subtraction and division invert the
order of the arguments before performing the operation: with `1`,
`2` and `3` in the stack, when you type `-` it will pop the values
`3` and `2` and push the result of `2 - 3`. This is in the tradition
of other postfix calculators and programming languages.
.
A description of the available commands follows.
.
.Ss Arithmetic operations
.
.Bl -tag -width Fl
.It Ic +
Pop two values `a` and `b` and push the result of `a + b`.
.
.It Ic -
Pop two values `a` and `b` and push the result of `b - a`.
.
.It Ic *
Pop two values `a` and `b` and push the result of `a * b`.
.
.It Ic /
Pop two values `a` and `b` and push the result of `b / a`.
.El
.
.Ss Modulo operation
.
.Bl -tag -width Fl
.It Ic %
Pop two values `a` and `b` and push the remainder of the Euclidean
division of `b` by `a`.
.El
.
.Ss Exponentiation
.
.Bl -tag -width Fl
.It Ic ^
Pop two values `a` and `b` and push the result of `b ^ a`.
.It Ic exp
Pop the value `a` and push the result of `e ^ a`.
.El
.
.Ss Logarithm
.
.Bl -tag -width Fl
.It Ic ln
Pop the value `a` and push its natural logarithm.
.It Ic log
Pop the value `a` and push its logarithm to base 10.
.El
.
.Ss Factorial
.
.Bl -tag -width Fl
.It Ic !
Pop the value `a` and push its factorial.
.El
.
.Ss Trigonometry
.
.Bl -tag -width Fl
.It Ic sin
Pop the value `a` and push its sine.
.It Ic cos
Pop the value `a` and push its cosine.
.It Ic tan
Pop the value `a` and push its tangent.
.It Ic asin
Pop the value `a` and push its arc sine.
.It Ic acos
Pop the value `a` and push its arc cosine.
.It Ic atan
Pop the value `a` and push its arc tangent.
.It Ic atan2
Pop two values `a` and `b` and push the arc tangent of `b / a`,
using the signs of `a` and `b` to determine the quadrant.
.El
.
.Ss Error function
.
.Bl -tag -width Fl
.It Ic erf
Pop the value `a` and push its error function.
.El
.
.Ss Summation
.
.Bl -tag -width Fl
.It Ic sum
Pop all the values in the stack and push their sum.
.It Ic add
Pop the value `a` and remove that many items from the stack. Push
their sum.
.El
.
.Ss Product
.
.Bl -tag -width Fl
.It Ic prod
Pop all the values in the stack and push their product.
.It Ic mul
Pop the value `a` and remove that many items from the stack. Push
their product.
.El
.
.Ss Rounding
.
.Bl -tag -width Fl
.It Ic ceil
Pop the value `a` and push the smallest integer value greater than or
equal to `a`.
.It Ic floor
Pop the value `a` and push the largest integer value less than or
equal to `a`.
.It Ic round
Pop the value `a` and push the integer value closest to `a`.
.El
.
.Ss Absolute value
.
.Bl -tag -width Fl
.It Ic abs
Pop the value `a` and push the non-negative value of `a`.
.El
.
.Ss Binary operations
.
.Bl -tag -width Fl
.It Ic and
Pop two values `a` and `b` and push the binary AND of `a` and `b`.
.It Ic or
Pop two values `a` and `b` and push the binary OR of `a` and `b`.
.It Ic xor
Pop two values `a` and `b` and push the binary XOR of `a` and `b`.
.El
.
.Ss Stack manipulation
.
.Bl -tag -width Fl
.It Ic swap
Pop two values `a` and `b` and push the values `a`, `b`.
.It Ic dup
Pop the value `a` and push the values `a`, `a`.
.It Ic roll
Pop two values `a` and `b` and rotate `b` elements in the stack `a`
times.
.It Ic drop
Remove the top of the stack.
.It Ic clear
Remove all the elements in the stack.
.It Ic count
Push the number of items in the stack.
.It Ic _
Push on the stack the result of the last operation.
.El
.
.Ss Stashing
.
.Bl -tag -width Fl
.It Ic stash
Pop the value `a` and move that many items to the stash.
.It Ic fetch
Pop the value `a` and move that many items from the stash.
.It Ic .
Stash the top of the stack.
.It Ic ,
Fetch one stashed item.
.It Ic :
Stash all the items in the stack.
.It Ic ;
Fetch all stashed items.
.El
.
.Ss History
.
Use
.Aq Ic C-p
/
.Aq Ic C-n
or
.Aq Ic Up
/
.Aq Ic Down
to navigate the history.
.
.Ss User defined operations
.
It is possible to define operations (or
.Em words ,
as they are usually called in stack based programming languages)
by editing the
.Pa words
configuration file. It is not created by default, but clac will use
some environment variables in order to search for word definitions.
.
.Bl -tag -width X
.It Ic $CLAC_WORDS
If set, it should point to a file containing word definitions.
.It Ic $XDG_CONFIG_HOME
If set, clac will search for
.Pa $XDG_CONFIG_HOME/clac/words
.It Ic $HOME
If set, clac will search for
.Pa $HOME/.config/clac/words
.El
.
.Ss How to define words
.
Words are defined as aliases, with one alias on each line. Empty
lines are ignored. Here are some examples:
.Pp
.Dl Sy pi No 3.14159265358979323846
.Dl Sy tau Qq "pi 2 *"
.Dl Sy sqrt Qq "0.5 ^"
.Pp
Note that an alias has two parts: a word to be defined and its
meaning. That's why the
.Sy tau
and
.Sy sqrt
definitions are enclosed in double quotes. If the double quotes are
removed, clac will complain that it can't parse the command. For
example, if we remove the double quotes from
.Sy sqrt
and start clac, we will get this error message:
.Pp
.Dl Incorrect definition: sqrt 0.5 ^
.Dl (~/.config/clac/words:3)
.Pp
User defined words can be used as if they were built-in commands:
.Pp
.Dl $ clac Qq "42 dup * pi *"
.Dl Sy 5541.76944093239
.
.Ss Comments
.
Any lines that begin with
.Sy #
are considered comments and are ignored. There are no inline comments,
and any
.Sy #
characters that are not the first character of a line are interpreted
literally. To define
.Sy #
as a word, wrap it in double quotes in the definition:
.Pp
.Dl # This is a comment.
.Dl \(dq#\(dq count
.Pp
The first line is ignored, and the second line assigns
.Sy #
to the operation
.Sy count.
.
.Ss How to list defined words
.
If you type `words` and hit enter, clac will list the defined words.
.
.Ss How to reload defined words
.
If you type `reload` and hit enter, clac will reload the words file.
.
.Sh EXAMPLES
.
While the most interesting aspect of clac is the ability to visualize
the stack as it is updated with each key press, at some point you
may want use clac just to get a quick result or call it from a
script. For that reason, clac can be used in non-interactive mode
by invoking it with an argument.
.Pp
Here are some examples of non-interactive invocations:
.Pp
.Dl $ clac Qq "3 4 +"
.Dl Sy 7
.Pp
.Dl $ clac Qq "2 3 4 +"
.Dl Sy 7
.Dl Sy 2
.Pp
When clac finishes evaluating the expression "2 3 4 +", there are
two elements in the stack: the number 7 at the top of the stack and
the number 2 at the bottom of the stack. The elements are printed
in order, one per line, starting from the top of the stack.
.Pp
This other example uses the stashing features. Let's say we want
to push two numbers and get the result of their multiplication plus
the square of the second number.
.Pp
.Dl $ clac Qq "4 3 dup dup * . * , +"
.Dl Sy 21
.Pp
Another example that uses the stash would be to get the average of
all the elements in the stack:
.Pp
.Dl $ clac Qq "1 2 3 4 count . sum , /"
.Dl Sy 2.5
.Pp
In fact, if you find yourself calculating averages very often, you can
define the word
.Sy avg
as
.Qq Sy "count . sum , /" .
.
.Sh AUTHOR
.An Michel Martens Aq mail@soveran.com


clac.c
/*
 * Copyright (c) 2017, Michel Martens <mail at soveran dot com>
 *
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are
 * met:
 *
 *  *  Redistributions of source code must retain the above copyright
 *     notice, this list of conditions and the following disclaimer.
 *
 *  *  Redistributions in binary form must reproduce the above copyright
 *     notice, this list of conditions and the following disclaimer in the
 *     documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
 * HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>
#include <errno.h>
#include <math.h>
#include <sys/stat.h>
#include <sys/types.h>
#include "linenoise.h"
#include "sds.h"

/* UI */
#define HINT_COLOR 33
#define NUMBER_FMT "%.15g"
#define OUTPUT_FMT "\x1b[33m= " NUMBER_FMT "\x1b[0m\n"
#define WORDEF_FMT "%s \x1b[33m\"%s\"\x1b[0m\n"

/* Config */
#define BUFFER_MAX 1024
#define HIST_DIR   "clac"
#define HIST_FILE  "clac/history"
#define WORDS_FILE "clac/words"
#define CAPACITY   0xFF

/* Stack */
#define count(S)   ((S)->top)
#define clear(S)   ((S)->top = 0)
#define isfull(S)  ((S)->top == CAPACITY)
#define isempty(S) ((S)->top == 0)

/* Arithmetic */
#define modulo(A, B) ((A) - (B) * floor((A) / (B)))

typedef struct stack {
	double items[CAPACITY];
	int top;
} stack;

typedef struct node {
	sds word;
	sds meaning;
	struct node *next;
} node;

static stack stacks[] = {{{}, 0}, {{}, 0}};
static stack *s0 = &stacks[0];
static stack *s1 = &stacks[1];
static node *head = NULL;
static node *tail = NULL;
static sds result;
static double hole = 0;

static int isoverflow(stack *s) {
	if (isfull(s)) {
		fprintf(stderr, "\r\nStack is full!\n");
		return 1;
	}

	return 0;
}

static double peek(stack *s) {
	if (isempty(s)) {
		return 0;
	}

	return s->items[s->top-1];
}

static void push(stack *s, double value) {
	if (!isoverflow(s)) {
		s->items[s->top++] = value;
	}
}

static double pop(stack *s) {
	if (isempty(s)) {
		return 0;
	}

	return s->items[--s->top];
}

static void move(stack *s, stack *t, int n) {
	while (!isempty(s) && n > 0) {
		push(t, pop(s));
		n--;
	}
}

static void roll(stack *s, stack *aux, int m, int n) {
	if (m > count(s)) {
		m = count(s);
	}

	if (m < 2) {
		return;
	}

	if (n < 0) {
		n = m - modulo(abs(n), m);
	}

	if (n == 0 || n == m) {
		return;
	}

	m--;

	double a;

	while (n > 0) {
		a = pop(s);
		move(s, aux, m);
		push(s, a);
		move(aux, s, m);
		n--;
	}
}

static double add(stack *s, int n) {
	double a = pop(s);

	while (!isempty(s) && n > 1) {
		a += pop(s);
		n--;
	}

	return a;
}

static double mul(stack *s, int n) {
	double a = pop(s);

	while (!isempty(s) && n > 1) {
		a *= pop(s);
		n--;
	}

	return a;
}

static node *get(sds word) {
	node *curr = head;

	while (curr != NULL) {
		if (!strcasecmp(word, curr->word)) {
			return curr;
		}

		curr = curr->next;
	}

	return NULL;
}

static void set(sds word, sds meaning) {
	node *curr = get(word);

	if (curr != NULL) {
		fprintf(stderr, "Duplicate definition of \"%s\"\n", word);
		curr->meaning = meaning;
		return;
	}

	curr = (node *) malloc(sizeof(node));

	if (curr == NULL) {
		fprintf(stderr, "Not enough memory to load words\n");
		exit(1);
	}

	curr->word = word;
	curr->meaning = meaning;
	curr->next = NULL;
	if (head == NULL) {
		head = curr;
	} else {
		tail->next = curr;
	}
	tail = curr;
}

static void cleanup() {
	node *curr;

	while (head != NULL) {
		curr = head;
		head = curr->next;

		sdsfree(curr->word);
		sdsfree(curr->meaning);

		free(curr);
	}
}

static int parse(sds input) {
	int argc;
	sds *argv = sdssplitargs(input, &argc);

	if (argc == 0) {
		sdsfreesplitres(argv, argc);
		return 0;
	}

	if (argc != 2) {
		sdsfreesplitres(argv, argc);
		fprintf(stderr, "Incorrect definition: %s\n", input);
		return 1;
	}

	set(argv[0], argv[1]);

	return 0;
}

static void load(sds filename) {
	FILE *fp;

	if ((fp = fopen(filename, "r")) == NULL) {
		if (errno == ENOENT) {
			return;
		}

		fprintf(stderr, "Can't open file %s\n", filename);
		sdsfree(filename);
		exit(1);
	}

	char buf[BUFFER_MAX+1];
	int linecount, i;

	sds *lines;
	sds content = sdsempty();

	while(fgets(buf, BUFFER_MAX+1, fp) != NULL) {
		content = sdscat(content, buf);
	}

	fclose(fp);

	lines = sdssplitlen(content, strlen(content), "\n", 1, &linecount);

	for (i = 0; i < linecount; i++) {
		lines[i] = sdstrim(lines[i], " \t\r\n");

		if (lines[i][0] == '#') {
		  continue;
		}

		if (parse(lines[i]) != 0) {
			sdsfreesplitres(lines, linecount);

			fprintf(stderr, "(%s:%d)\n", filename, i+1);
			exit(1);
		}
	}

	sdsfreesplitres(lines, linecount);
	sdsfree(content);
}

static void eval(const char *input);

static void process(sds word) {
	double a, b;
	char *z;
	node *n;

	if (!strcmp(word, "_")) {
		push(s0, hole);
	} else if (!strcmp(word, "+")) {
		if (count(s0) > 1) {
			a = pop(s0);
			b = pop(s0);
			push(s0, a + b);
		}
	} else if (!strcmp(word, "-")) {
		if (count(s0) > 1) {
			a = pop(s0);
			b = pop(s0);
			push(s0, b - a);
		}
	} else if (!strcmp(word, "*")) {
		if (count(s0) > 1) {
			a = pop(s0);
			b = pop(s0);
			push(s0, b * a);
		}
	} else if (!strcmp(word, "/")) {
		if (count(s0) > 1) {
			a = pop(s0);
			b = pop(s0);
			push(s0, b / a);
		}
	} else if (!strcmp(word, "%")) {
		if (count(s0) > 1) {
			a = pop(s0);
			b = pop(s0);
			push(s0, modulo(b, a));
		}
	} else if (!strcmp(word, "^")) {
		if (count(s0) > 1) {
			a = pop(s0);
			b = pop(s0);
			push(s0, pow(b, a));
		}
	} else if (!strcmp(word, "or")) {
		if (count(s0) > 1) {
			a = pop(s0);
			b = pop(s0);
			push(s0, (int)fabs(b)|(int)fabs(a));
		}
	} else if (!strcmp(word, "and")) {
		if (count(s0) > 1) {
			a = pop(s0);
			b = pop(s0);
			push(s0, (int)fabs(b)&(int)fabs(a));
		}
	} else if (!strcmp(word, "xor")) {
		if (count(s0) > 1) {
			a = pop(s0);
			b = pop(s0);
			push(s0, (int)fabs(b)^(int)fabs(a));
		}
	} else if (!strcasecmp(word, "sum")) {
		push(s0, add(s0, count(s0)));
	} else if (!strcasecmp(word, "add")) {
		push(s0, add(s0, pop(s0)));
	} else if (!strcasecmp(word, "prod")) {
		push(s0, mul(s0, count(s0)));
	} else if (!strcasecmp(word, "mul")) {
		push(s0, mul(s0, pop(s0)));
	} else if (!strcasecmp(word, "abs")) {
		if (count(s0) > 0) {
			push(s0, fabs(pop(s0)));
		}
	} else if (!strcasecmp(word, "ceil")) {
		if (count(s0) > 0) {
			push(s0, ceil(pop(s0)));
		}
	} else if (!strcasecmp(word, "floor")) {
		if (count(s0) > 0) {
			push(s0, floor(pop(s0)));
		}
	} else if (!strcasecmp(word, "round")) {
		if (count(s0) > 0) {
			push(s0, round(pop(s0)));
		}
	} else if (!strcasecmp(word, "sin")) {
		if (count(s0) > 0) {
			push(s0, sin(pop(s0)));
		}
	} else if (!strcasecmp(word, "cos")) {
		if (count(s0) > 0) {
			push(s0, cos(pop(s0)));
		}
	} else if (!strcasecmp(word, "tan")) {
		if (count(s0) > 0) {
			push(s0, tan(pop(s0)));
		}
	} else if (!strcasecmp(word, "asin")) {
		if (count(s0) > 0) {
			push(s0, asin(pop(s0)));
		}
	} else if (!strcasecmp(word, "acos")) {
		if (count(s0) > 0) {
			push(s0, acos(pop(s0)));
		}
	} else if (!strcasecmp(word, "atan")) {
		if (count(s0) > 0) {
			push(s0, atan(pop(s0)));
		}
	} else if (!strcasecmp(word, "atan2")) {
		if (count(s0) > 1) {
			a = pop(s0);
			b = pop(s0);
			push(s0, atan2(b, a));
		}
	} else if (!strcasecmp(word, "ln")) {
		if (count(s0) > 0) {
			push(s0, log(pop(s0)));
		}
	} else if (!strcasecmp(word, "log")) {
		if (count(s0) > 0) {
			push(s0, log10(pop(s0)));
		}
	} else if (!strcasecmp(word, "exp")) {
		if (count(s0) > 0) {
			push(s0, exp(pop(s0)));
		}
	} else if (!strcasecmp(word, "erf")) {
		if (count(s0) > 0) {
			push(s0, erf(pop(s0)));
		}
	} else if (!strcasecmp(word, "!")) {
		if (count(s0) > 0) {
			a = pop(s0);

			if (a == 0) {
				push(s0, 1);
			} else {
				push(s0, a * tgamma(a));
			}
		}
	} else if (!strcasecmp(word, "dup")) {
		if (!isempty(s0)) {
			push(s0, peek(s0));
		}
	} else if (!strcasecmp(word, "roll")) {
		a = pop(s0);
		b = pop(s0);

		roll(s0, s1, b, a);
	} else if (!strcasecmp(word, "swap")) {
		if (count(s0) > 1) {
			a = pop(s0);
			b = pop(s0);

			push(s0, a);
			push(s0, b);
		}
	} else if (!strcasecmp(word, "drop")) {
		pop(s0);
	} else if (!strcasecmp(word, "count")) {
		push(s0, (double) count(s0));
	} else if (!strcasecmp(word, "clear")) {
		clear(s0);
	} else if (!strcasecmp(word, "stash")) {
		move(s0, s1, pop(s0));
	} else if (!strcasecmp(word, "fetch")) {
		move(s1, s0, pop(s0));
	} else if (!strcasecmp(word, ".")) {
		move(s0, s1, 1);
	} else if (!strcasecmp(word, ",")) {
		move(s1, s0, 1);
	} else if (!strcasecmp(word, ":")) {
		move(s0, s1, count(s0));
	} else if (!strcasecmp(word, ";")) {
		move(s1, s0, count(s1));
	} else if ((n = get(word)) != NULL) {
		eval(n->meaning);
	} else {
		a = strtod(word, &z);

		if (*z == '\0') {
			push(s0, a);
		} else if (!isalpha(word[0])) {
			push(s0, NAN);
		}
	}
}

static void eval(const char *input) {
	int i, argc;

	sds *argv = sdssplitargs(input, &argc);

	for (i = 0; i < argc; i++) {
		process(argv[i]);
	}

	sdsfreesplitres(argv, argc);
}

static void completion(const char *input, linenoiseCompletions *lc) {}

static char *hints(const char *input, int *color, int *bold) {
	int i;

	clear(s0);
	clear(s1);

	eval(input);
	sdsclear(result);

	result = sdscat(result, " ");

	for (i = 0; i < count(s0); i++) {
		result = sdscatprintf(result, " " NUMBER_FMT, s0->items[i]);
	}

	if (!isempty(s1)) {
		result = sdscat(result, " ⋮");

		for (i = s1->top-1; i > -1; i--) {
			result = sdscatprintf(result, " " NUMBER_FMT, s1->items[i]);
		}
	}

	*color = HINT_COLOR;

	return result;
}

static sds buildpath(const char *fmt, const char *dir, const char *file) {
	return sdscatfmt(sdsempty(), fmt, dir, file);
}

static void config() {
	sds filename = NULL;

	if (getenv("CLAC_WORDS") != NULL) {
		filename = sdsnew(getenv("CLAC_WORDS"));
	} else if (getenv("XDG_CONFIG_HOME") != NULL) {
		filename = buildpath("%s/%s", getenv("XDG_CONFIG_HOME"), WORDS_FILE);
	} else if (getenv("HOME") != NULL) {
		filename = buildpath("%s/.config/%s", getenv("HOME"), WORDS_FILE);
	}

	if (filename) {
		load(filename);
		sdsfree(filename);
	}
}

static sds history_config() {
	sds directory = NULL;
	sds filename = NULL;

	if (getenv("XDG_DATA_HOME") != NULL) {
		directory = buildpath("%s/%s", getenv("XDG_DATA_HOME"), HIST_DIR);
		filename = buildpath("%s/%s", getenv("XDG_DATA_HOME"), HIST_FILE);
	} else if (getenv("HOME") != NULL) {
		directory = buildpath("%s/.local/share/%s", getenv("HOME"), HIST_DIR);
		filename = buildpath("%s/.local/share/%s", getenv("HOME"), HIST_FILE);
	} else {
		return NULL;
	}

	mkdir(directory, 0700);
	sdsfree(directory);

	return filename;
}

int main(int argc, char **argv) {
	char *line;

	result = sdsempty();

	config();

	if (argc == 2) {
		eval(argv[1]);

		while (count(s0) > 0) {
			printf(NUMBER_FMT "\n", pop(s0));
		}

		exit(0);
	}

	if (argc > 2) {
		fprintf(stderr, "usage: clac [expression]\n");
		exit(1);
	}

	sds history = history_config();

	linenoiseSetHintsCallback(hints);
	linenoiseSetCompletionCallback(completion);

	if (history) {
		linenoiseHistoryLoad(history);
	}

	while((line = linenoise("> ")) != NULL) {
		if (!strcmp(line, "words")) {
			node *curr = head;

			while (curr != NULL) {
				printf(WORDEF_FMT, curr->word, curr->meaning);
				curr = curr->next;
			}
		} else if (!strcmp(line, "reload")) {
			cleanup();
			config();
		} else if (!isempty(s0)) {
			hole = peek(s0);
			clear(s0);
			printf(OUTPUT_FMT, hole);
		}

		sdsclear(result);
		linenoiseHistoryAdd(line);
		free(line);

		if (history) {
			linenoiseHistorySave(history);
		}
	}

	sdsfree(result);
	sdsfree(history);
	cleanup();

	return 0;
}
